slacksocketmode/socket_mode_managed_conn.go

```diff
@@ -0,0 +1,469 @@
+	ticker := time.NewTicker(smc.pingInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		// catch "stop" signal on channel close
+		case intentional := <-smc.killChannel:
+			_ = smc.killConnection(intentional, errorsx.String("signaled"))
+			return
+		// detect when the connection is dead.
+		case <-smc.pingDeadman.C:
+			_ = smc.killConnection(false, slack.ErrRTMDeadman)
+			return
+		// listen for messages that need to be sent
+		case msg := <-smc.outgoingMessages:
+			smc.sendOutgoingMessage(msg)
+			// listen for incoming messages that need to be parsed
+		case rawEvent := <-events:
+			_ = smc.handleRawEvent(rawEvent)
+		}
+	}
+}
+
+// handleIncomingEvents monitors the RTM's opened websocket for any incoming
+// events. It pushes the raw events into the channel.
+//
+// This will stop executing once the RTM's when a fatal error is detected, or
+// a disconnect occurs.
+func (smc *Client) handleIncomingEvents(events chan json.RawMessage) {
+	for {
+		if err := smc.receiveIncomingEvent(events); err != nil {
+			select {
+			case smc.killChannel <- false:
+			case <-smc.disconnected:
+			}
+			return
+		}
+	}
+}
+
+func (smc *Client) sendWithDeadline(msg interface{}) error {
+	// set a write deadline on the connection
+	if err := smc.conn.SetWriteDeadline(time.Now().Add(10 * time.Second)); err != nil {
+		return err
+	}
+	if err := smc.conn.WriteJSON(msg); err != nil {
+		return err
+	}
+	// remove write deadline
+	return smc.conn.SetWriteDeadline(time.Time{})
+}
+
+func (smc *Client) internalEvent(tpe string, data interface{}) SocketModeEvent {
+	return SocketModeEvent{Type: tpe, Data: data}
+}
+
+func (smc *Client) externalEvent(tpe string, data interface{}) SocketModeEvent {
+	return SocketModeEvent{Type: tpe, Data: data}
+}
+
+// sendOutgoingMessage sends the given OutgoingMessage to the slack websocket.
+//
+// It does not currently detect if a outgoing message fails due to a disconnect
+// and instead lets a future failed 'PING' detect the failed connection.
+func (smc *Client) sendOutgoingMessage(msg slack.OutgoingMessage) {
+	smc.Debugln("Sending message:", msg)
+	if len([]rune(msg.Text)) > slack.MaxMessageTextLength {
+		smc.IncomingEvents <- smc.internalEvent("outgoing_error", &slack.MessageTooLongEvent{
+			Message:   msg,
+			MaxLength: slack.MaxMessageTextLength,
+		})
+		return
+	}
+
+	if err := smc.sendWithDeadline(msg); err != nil {
+		smc.IncomingEvents <- smc.internalEvent("outgoing_error", &slack.OutgoingErrorEvent{
+			Message:  msg,
+			ErrorObj: err,
+		})
+	}
+}
+
+// ack tells Slack that the we have received the SocketModeRequest denoted by the envelope ID,
+// by sending back the envelope ID over the WebSocket connection.
+func (smc *Client) ack(envelopeID string) error {
+	smc.Debugln("Sending ACK ", envelopeID)
+
+	// See https://github.com/slackapi/node-slack-sdk/blob/c3f4d7109062a0356fb765d53794b7b5f6b3b5ae/packages/socket-mode/src/SocketModeClient.ts#L417
+	msg := map[string]interface{}{"envelope_id": envelopeID}
```

Comment:
The response message can have "payload" in addition to "envelop_id" for interactive / slash_commands types, while events_api requires only "envelope_id".

* https://api.slack.com/apis/connections/socket-implement#acknowledge
* https://github.com/slackapi/java-slack-sdk/tree/v1.5.1/slack-api-client/src/main/java/com/slack/api/socket_mode/response

`slack.OutgoingMessage` is not compatible with Socket Mode.

Reply:
Ah thanks! First of all, I didn't realize till now that `slash_commands` and `interactive` are top-level types that are comparable to `events_api` and `hello` and so on.

And it's really good to know that the way we ack the socket mode "request" (or event, which do you prefer? I thought you call it `SocketModeRequest` in python-sdk and `event` in node-sdk btw)

I will soon try modifying `examples/socketmode` and the library and test it accordingly and circle back to you.

Reply:
The first changeset for this: https://github.com/slack-go/slack/pull/885/commits/88722e9b1ed603fbea5d962aa0dd9a0f0b7cc639#diff-f5d9e76c79011a908909fa0f14332b25faba7ef70c64f12682937d35382f3f47R395-R425